zookeeper-docs/src/main/resources/markdown/zookeeperTools.md

```diff
@@ -205,6 +206,210 @@ USAGE: SnapshotFormatter [-d|-json] snapshot_file
+ -n,--nodes <NODETHRESHOLD>   (Required) The descendant node delta size threshold, in nodes, for printing the node.
+ -r,--right <RIGHT>           (Required) The right snapshot file.
+```
+Example Command:
+
+```
+./bin/zkSnapshotComparer.sh -l /zookeeper-data/backup/snapshot.d.snappy -r /zookeeper-data/backup/snapshot4 -b 2 -n 1
+```
+
+Example Output:
+```
+...
+Deserialized snapshot in snapshot.d.snappy in 0.043911 seconds
+Processed data tree in 0.075437 seconds
+2020-02-02 23:53:27,931 [myid:] - INFO  [main:WatchManagerFactory@42] - Using org.apache.zookeeper.server.watch.WatchManager as watch manager
+2020-02-02 23:53:27,931 [myid:] - INFO  [main:WatchManagerFactory@42] - Using org.apache.zookeeper.server.watch.WatchManager as watch manager
+Deserialized snapshot in snapshot.44 in 0.002155 seconds
+Processed data tree in 0.000389 seconds
+Node count: 10
+Total size: 0
+Max depth: 4
+Count of nodes at depth 0: 1
+Count of nodes at depth 1: 2
+Count of nodes at depth 2: 4
+Count of nodes at depth 3: 3
+
+Node count: 22
+Total size: 2903
+Max depth: 5
+Count of nodes at depth 0: 1
+Count of nodes at depth 1: 2
+Count of nodes at depth 2: 4
+Count of nodes at depth 3: 7
+Count of nodes at depth 4: 8
+
+Printing analysis for nodes difference larger than 2 bytes or node count difference larger than 1.
+Analysis for depth 0
+Node  found in both trees. Delta: 2903 bytes, 12 descendants
+Analysis for depth 1
+Node /zk_test found in both trees. Delta: 2903 bytes, 12 descendants
+Analysis for depth 2
+Node /zk_test/gz found in both trees. Delta: 730 bytes, 3 descendants
+Node /zk_test/snappy found in both trees. Delta: 2173 bytes, 9 descendants
+Analysis for depth 3
+Node /zk_test/gz/12345 found in both trees. Delta: 9 bytes, 1 descendants
+Node /zk_test/gz/a found only in right tree. Descendant size: 721. Descendant count: 0
+Node /zk_test/snappy/anotherTest found in both trees. Delta: 1738 bytes, 2 descendants
+Node /zk_test/snappy/test_1 found only in right tree. Descendant size: 344. Descendant count: 3
+Node /zk_test/snappy/test_2 found only in right tree. Descendant size: 91. Descendant count: 2
+Analysis for depth 4
+Node /zk_test/gz/12345/abcdef found only in right tree. Descendant size: 9. Descendant count: 0
+Node /zk_test/snappy/anotherTest/abc found only in right tree. Descendant size: 1738. Descendant count: 0
+Node /zk_test/snappy/test_1/a found only in right tree. Descendant size: 93. Descendant count: 0
+Node /zk_test/snappy/test_1/b found only in right tree. Descendant size: 251. Descendant count: 0
+Node /zk_test/snappy/test_2/xyz found only in right tree. Descendant size: 33. Descendant count: 0
+Node /zk_test/snappy/test_2/y found only in right tree. Descendant size: 58. Descendant count: 0
+All layers compared.s
+```
+
+#### Interactive Mode
+Use "-i" or "--interactive" to enter interactive mode:
+```
+./bin/zkSnapshotComparer.sh -l /zookeeper-data/backup/snapshot.d.snappy -r /zookeeper-data/backup/snapshot.44 -b 2 -n 1 -i
+```
```

Comment:
For `interactive` mode:
- When pressing `Enter` twice: I got this:
```
Exception in thread "main" java.util.IllegalFormatConversionException: d != java.lang.String
	at java.base/java.util.Formatter$FormatSpecifier.failConversion(Formatter.java:4426)
	at java.base/java.util.Formatter$FormatSpecifier.printInteger(Formatter.java:2938)
	at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:2892)
	at java.base/java.util.Formatter.format(Formatter.java:2673)
	at java.base/java.util.Formatter.format(Formatter.java:2609)
	at java.base/java.lang.String.format(String.java:2897)
	at org.apache.zookeeper.server.SnapshotComparer.printLeftOnly(SnapshotComparer.java:428)
	at org.apache.zookeeper.server.SnapshotComparer.compareNodes(SnapshotComparer.java:392)
	at org.apache.zookeeper.server.SnapshotComparer.compareLine(SnapshotComparer.java:359)
	at org.apache.zookeeper.server.SnapshotComparer.compareTrees(SnapshotComparer.java:306)
	at org.apache.zookeeper.server.SnapshotComparer.compareSnapshots(SnapshotComparer.java:148)
	at org.apache.zookeeper.server.SnapshotComparer.main(SnapshotComparer.java:121)
```

- 123456788999999999999999999 is not path, it is a corner case number I want to test?
```
Current depth is 0
Press enter to move to print current depth layer;
Type a number to jump to and print all nodes at a given depth;
Enter an ABSOLUTE path to print the immediate subtree of a node.
123456788999999999999999999
Printing analysis for nodes difference larger than 2 bytes or node count difference larger than 1.
Analysis for node 123456788999999999999999999
Path 123456788999999999999999999 is neither found in left tree nor right tree.
Current depth is 0
```
- add a category for this?
```
  1: Press enter to move to print current depth layer;
  2: Type a number to jump to and print all nodes at a given depth;
  3: Enter an ABSOLUTE path to print the immediate subtree of a node.
```

- add a space between the results. more clear?
```
Press enter to move to print current depth layer;
Type a number to jump to and print all nodes at a given depth;
Enter an ABSOLUTE path to print the immediate subtree of a node.
/zookeeper
Printing analysis for nodes difference larger than 2 bytes or node count difference larger than 1.
Analysis for node /zookeeper
Node /zookeeper/config found in both trees. Delta: -154 bytes, 0 descendants
Filtered node /zookeeper/quota of left size 31, right size 31
Current depth is 0

Press enter to move to print current depth layer;
Type a number to jump to and print all nodes at a given depth;
Enter an ABSOLUTE path to print the immediate subtree of a node.
3
Printing analysis for nodes difference larger than 2 bytes or node count difference larger than 1.
Analysis for depth 3
Filtered node /a/b/c of left size 6, right size 6
```

Reply:
Great catch! Added fix/improvements to address the above in the latest revision. I used `-` instead of numbering to differentiate each option, in case users mistakenly think they should enter a number to proceed.